Describe the points at which ini files are processed

kpinc opened the pull request.

stevepiercy commented on docs/narr/project.rst:
```diff
@@ -856,12 +860,19 @@ also informs Python that the directory which contains it is a *package*.
-   specify renderers with the ``.jinja2`` extension.
+   specify renderers with the ``.jinja2`` extension.  The expectation
+   is that the extension applies what is relevant in the ``.ini`` file
+   at this point.  Consult your extension's documentation.
```
s/extension/add-on in both new sentences. `.jinja2` is a file extension (so the usage in the original sentence is correct in that context), whereas pyramid_jinja2 is the add-on that applies ini stuff and has documentation.

stevepiercy commented on docs/narr/project.rst:
```diff
@@ -856,12 +860,19 @@ also informs Python that the directory which contains it is a *package*.
-   Within this function, application configuration is performed.
+   :term:`PasteDeploy` configures logging from the ``.ini`` file
+   before ``main`` is called.  Within ``main`` further application
+   configuration is performed.
```
This flow is awkward. We talk about `main`, then jump to PasteDeploy `ini`, then back to `main`.

Please take a look at lines 658-666. I think that it would be better to insert a subsection heading up there, like so:

```
.. index:: 
    single: ini file processing
    pair: extensions; configuration
    pair: logging; configuration
    pair: PasteDeploy; configuration

.. _myproject_ini_logging:

Logging configuration
^^^^^^^^^^^^^^^^^^^^^

The sections after ``# logging configuration`` represent Python's standard
```

Then refer to that, like so:

```
Within this function, application configuration is performed, except for :ref:`logging as described previously <myproject_ini_logging>`.
```

p.s.: Narrative docs do not need to be hard-wrapped to 79 columns. It is [preferred not to wrap sentences, and to put each sentence on its own line](https://docs.pylonsproject.org/projects/docs-style-guide/#use-of-whitespace). This makes diffing narrative docs much easier because whitespace noise from wrapping is eliminated.

stevepiercy commented on docs/narr/project.rst:
```diff
@@ -856,12 +860,19 @@ also informs Python that the directory which contains it is a *package*.
    WSGI application.  This function is meant to be called by the
    :term:`PasteDeploy` framework as a result of running ``pserve``.
```
I think this will help.

```
``pserve`` by virtue of the ``[server:main]`` block in the specified ``.ini`` file.
```

kpinc commented:
Hi Steve,

On Mon, 29 Oct 2018 13:11:12 -0700
Steve Piercy <notifications@github.com> wrote:

> I'm not certain these requested changes will improve the docs. I'd
> like at least one more core developer to help on this.

Thanks very much for the review.  I've made large-ish changes.
You should see a new patch in the pull request (if I've
operated github correctly).

>     Line 8 adds support for Jinja2 templating bindings, allowing us to
> -   specify renderers with the ``.jinja2`` extension.
> +   specify renderers with the ``.jinja2`` extension.  The expectation
> +   is that the extension applies what is relevant in the ``.ini``
> file
> +   at this point.  Consult your extension's documentation.
> 
> s/extension/add-on in both new sentences. `.jinja2` is a file
> extension (so the usage in the original sentence is correct in that
> context), whereas pyramid_jinja2 is the add-on that applies ini stuff
> and has documentation.

Fixed.

> > @@ -856,12 +860,19 @@ also informs Python that the directory which
> > contains it is a *package*.  
>     WSGI application.  This function is meant to be called by the
>     :term:`PasteDeploy` framework as a result of running ``pserve``.
>  
> -   Within this function, application configuration is performed.
> +   :term:`PasteDeploy` configures logging from the ``.ini`` file
> +   before ``main`` is called.  Within ``main`` further application
> +   configuration is performed.
> 
> This flow is awkward. We talk about `main`, then jump to PasteDeploy
> `ini`, then back to `main`.

I believe I've fixed this by flipping the first sentences subject and
object.

> Please take a look at lines 658-666. I think that it would be better
> to insert a subsection heading up there, like so:
> 
> ```
> .. index:: 
>     single: ini file processing
>     pair: extensions; configuration
>     pair: logging; configuration
>     pair: PasteDeploy; configuration
> 
> .. _myproject_ini_logging:
> 
> Logging configuration
> ^^^^^^^^^^^^^^^^^^^^^
> 
> The sections after ``# logging configuration`` represent Python's
> standard ```
> 
> Then refer to that, like so:
> 
> ```
> Within this function, application configuration is performed, except
> for :ref:`logging as described previously <myproject_ini_logging>`.
> ```

The docs already link to <myproject_ini_logging> and there's no
sub-section now.  In the interest of keeping the docs "small"
I'm going to leave out having a sub-section heading.

I believe that the right place to point index entries which
have to do with logging configuration and the main function is
at the point in the docs where the main function is described.
When you're writing main you want to know not _how_ to configure
logging but when/whether the logging is configured with respect
to when main is executed.

> p.s.: Narrative docs do not need to be hard-wrapped to 79 columns. It
> is [preferred not to wrap sentences, and to put each sentence on its
> own
> line](https://docs.pylonsproject.org/projects/docs-style-guide/#use-of-whitespace).
> This makes diffing narrative docs much easier because whitespace
> noise from wrapping is eliminated.

Done.

> > @@ -856,12 +860,19 @@ also informs Python that the directory which
> > contains it is a *package*.  
>     WSGI application.  This function is meant to be called by the
>     :term:`PasteDeploy` framework as a result of running ``pserve``.
> 
> I think this will help.
> 
> ```
> ``pserve`` by virtue of the ``[server:main]`` block in the specified
> ``.ini`` file. ```

I didn't do this.  But I did take the gist of the advice you offer
here and use links to help shorten and clarify sentences.

Thanks for the help.

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

kpinc commented:
On Mon, 29 Oct 2018 13:11:12 -0700
Steve Piercy <notifications@github.com> wrote:

> Please take a look at lines 658-666. I think that it would be better
> to insert a subsection heading up there, like so:
> 
> ```
> .. index:: 
>     single: ini file processing
>     pair: extensions; configuration
>     pair: logging; configuration
>     pair: PasteDeploy; configuration
> 
> .. _myproject_ini_logging:
> 
> Logging configuration
> ^^^^^^^^^^^^^^^^^^^^^
> 
> The sections after ``# logging configuration`` represent Python's
> standard ```

I think if we were to insert a "Logging configuration"
sub-section heading we'd also want a sub-section for
[app:main] and another for [server:main].

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

kpinc commented:
On Tue, 30 Oct 2018 03:24:59 +0000 (UTC)
Steve Piercy <notifications@github.com> wrote:

> I made a suggestion that might help with that one awkward sentence.
> 
> >    
>     Line 7 opens a context manager with an instance of
> a :term:`Configurator`.
> +   This applies those :ref:`[app:main] <app_main>` lines of the
> ``.ini`` file understood by :app:`Pyramid`.
> +   E.g. ``pyramid.includes``.
> 
> This is an awkward sentence. Suggest two sentences.
> 
> ```
> This applies the lines of settings declared in the
> block :ref:`[app:main] <app_main>` in the specified ``.ini`` file.
> These settings are understood by :app:`Pyramid`, such as
> ``pyramid.includes = pyramid_debugtoolbar``. ```

I was shooting for brevity.  But that's no good if it's not clear.
As long as I've got to split the sentence in two I wanted to
add the detail that explains what it means to "apply"
an .ini file.  That the .ini file content is applied to
a configuration.  And I added that Configurator-s are what
make configurations, for extra clarity.

Hopefully this is not too much detail.  Overlong means
not read/internalized.  We could cut out the ", the configuration
returned by the Configurator".  This might improve it.

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

kpinc commented:
On Mon, 29 Oct 2018 22:59:53 -0500
"Karl O. Pinc" <kop@meme.com> wrote:

> On Tue, 30 Oct 2018 03:24:59 +0000 (UTC)
> Steve Piercy <notifications@github.com> wrote:

> > This is an awkward sentence. Suggest two sentences.
> > <snip>

> I was shooting for brevity. 

> Hopefully this is not too much detail.  Overlong means
> not read/internalized.  We could cut out the ", the configuration
> returned by the Configurator".  This might improve it.

To really make it brief, and add punch, we could change:

 This applies lines in the .ini file to config, the configuration
 returned by Configurator.

to just:

 This applies lines in the .ini file.

I don't know what the right balance is here.

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

kpinc commented:
On Mon, 29 Oct 2018 23:18:12 -0500
"Karl O. Pinc" <kop@meme.com> wrote:

> On Mon, 29 Oct 2018 22:59:53 -0500
> "Karl O. Pinc" <kop@meme.com> wrote:
> 
> > On Tue, 30 Oct 2018 03:24:59 +0000 (UTC)
> > Steve Piercy <notifications@github.com> wrote:  
> 
> > > This is an awkward sentence. Suggest two sentences.
> > > <snip>  
> 
> > I was shooting for brevity.   
> 
> > Hopefully this is not too much detail.  Overlong means
> > not read/internalized.  We could cut out the ", the configuration
> > returned by the Configurator".  This might improve it.  
> 
> To really make it brief, and add punch, we could change:
> 
>  This applies lines in the .ini file to config, the configuration
>  returned by Configurator.
> 
> to just:
> 
>  This applies lines in the .ini file.

I'm increasingly inclined to cut that sentence  
down to the above minimum.  Short is good.  
No distractions.  But I'm not      
going to submit any more changes without feedback from you.

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

stevepiercy commented:
@kpinc none of your commits since 11 hours ago have come through.

kpinc commented:
On Mon, 29 Oct 2018 22:58:53 -0700
Steve Piercy <notifications@github.com> wrote:

> @kpinc none of your commits since 11 hours ago have come through.

I am pushing with -f after rebase to my "iniprocessing" branch.
https://github.com/kpinc/pyramid/commit/5c51923c3afe273a1917e8f93c5733ec819e0a3d

The latest changes show for me on the pull request:

https://github.com/Pylons/pyramid/pull/3405/commits/5c51923c3afe273a1917e8f93c5733ec819e0a3d

But github does not seem to update the timestamp.

And the "conversation" in the hunks above and below
the hunk I changed seems to have disappeared.

...

Next time for a simple one-patch pull request like this I can
submit multiple commits for you as we make changes
to the pull request if that's what you like.

But I can't see how that would work if the pull request itself
is supposed to consist of multiple commits.  There would be
no way to recover the intended individual commits.  I think
I would still have to rebase the commits and push -f to
github.

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

mmerickel commented on docs/narr/project.rst:
```diff
@@ -856,12 +863,16 @@ also informs Python that the directory which contains it is a *package*.
-   specify renderers with the ``.jinja2`` extension.
+   Line 8 adds support for Jinja2 templating bindings, allowing us to specify renderers with the ``.jinja2`` extension.
+   The expectation is that the add-on applies what is relevant in the ``.ini`` file at this point.
+   Consult your add-on's documentation.
```
settings come from the settings field passed to the config via `Configurator(settings=settings)`. You choose to pass the ones from the ini, but I'm not comfortable just saying everywhere that things are from the ini file.

kpinc commented:
On Tue, 30 Oct 2018 08:29:51 -0700
Michael Merickel <notifications@github.com> wrote:

> settings come from the settings field passed to the config via
> `Configurator(settings=settings)`. You choose to pass the ones from
> the ini, but I'm not comfortable just saying everywhere that things
> are from the ini file.

I changed the whole bullet point to read:


Line 8 adds support for Jinja2 templating bindings. Settings are applied to the configuration allowing us to specify renderers with the .jinja2 extension. The source of these settings are add-on dependent.


While it should be obvious from the "include()" name of the function
call that the line alters configuration I think it's helpful
to be explicit about this.  Otherwise your eyes glaze over and
your brain goes "yeah, yeah, yeah".  Knowing that there are
actual settings behind the scenes here seems pertinent.

The last sentence "The source of these settings are add-on 
dependent." might be too much information at this point
because the whole idea of the page being an overview.

That's my thought process.  I see the point if you think that
any changes to this bullet point are too much.

Thanks for the review.

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

kpinc commented:
Hi,

I have started over here.  Reverted everything and began from the tip of master.

The basic idea is to make the project page smaller, adding links to
elsewhere when needed.  And possibly adding sentences to elsewhere
to clarify, notably the startup page.

kpinc commented on docs/narr/project.rst:
```diff
@@ -856,12 +863,16 @@ also informs Python that the directory which contains it is a *package*.
 
    Line 7 opens a context manager with an instance of a :term:`Configurator`.
+   This applies those :ref:`[app:main] <app_main>` lines of the ``.ini`` file understood by :app:`Pyramid`.
+   E.g. ``pyramid.includes``.
```
Started over based on mmerkel's comments.started

stevepiercy commented on docs/narr/advconfig.rst:
```diff
@@ -168,9 +168,11 @@ Using ``config.commit()``
 
 You can manually commit a configuration by using the
 :meth:`~pyramid.config.Configurator.commit` method between configuration calls.
+After a commit :term:`configuration declaration`\s which would otherwise result in a conflict may be added to a :term:`configurator` and will override previously committed directives.
```
Two shorter sentences for the English-challenged.

```
After a commit, more :term:`configuration declaration`\s may be added to a :term:`configurator`.
If any of these new :term:`configuration declaration`\s conflict with previously committed directives, they will override them.
```

stevepiercy commented on docs/narr/logging.rst:
```diff
@@ -32,6 +32,7 @@ you to send messages to :mod:`Python standard library logging package
 :term:`PasteDeploy` ``development.ini`` and ``production.ini`` files created
 when you use our cookiecutter include a basic configuration for the Python
 :mod:`logging` package.
+These ini file sections are passed to the `logging module's config file configuration engine <https://docs.python.org/3.7/howto/logging.html#configuring-logging>`_.
```
```
``ini`` [or ``.ini``, your choice]
```

stevepiercy commented on docs/narr/project.rst:
```diff
@@ -871,6 +869,8 @@ also informs Python that the directory which contains it is a *package*.
    Line 11 returns a :term:`WSGI` application to the caller of the function
    (Pyramid's pserve).
 
+   See :ref:`the_startup_process` for more about the application's :ref:`settings <startup_settings>` and ``main``\'s :ref:`arguments <startup_constructor_arguments>` and statements.
```
Let's make this a [`.. seealso::`](https://docs.pylonsproject.org/projects/docs-style-guide/#see-also) directive so it stands out from the flow of text.

stevepiercy commented on docs/narr/startup.rst:
```diff
@@ -107,9 +115,10 @@ Here's a high-level time-ordered overview of what happens when you press
-   dictionary captured via the ``**settings`` kwarg as its ``settings``
-   argument.
+   :class:`~pyramid.config.Configurator` instance.
+   It holds the application's settings and is able to :term:`commit` any :term:`configuration declaration`\s the settings contain because the dictionary captured via the ``**settings`` kwarg is passed to the :class:`~pyramid.config.Configurator` in its``settings`` argument.
```
Two sentences.

```
The instance holds the application's settings.
It is able to :term:`commit` any :term:`configuration declaration`\s contained in the settings because the dictionary captured via the ``**settings`` kwarg is passed to the :class:`~pyramid.config.Configurator` in its``settings`` argument.
```

kpinc commented:
On Mon, 19 Nov 2018 11:22:18 +0000 (UTC)
Steve Piercy <notifications@github.com> wrote:

> Breaking longer sentences into more, shorter sentences improves
> readability. 

Always good advice.  Thanks.

New patches added.  Again ready for review.

If you ever feel this whole thing is becoming an
unrelated blob of commits let me know and I can
split it up into separate pull requests.

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

stevepiercy commented on docs/narr/advconfig.rst:
```diff
@@ -168,9 +168,12 @@ Using ``config.commit()``
 
 You can manually commit a configuration by using the
 :meth:`~pyramid.config.Configurator.commit` method between configuration calls.
+After a commit more :term:`configuration declaration`\s may be added to a :term:`configurator`.
```
"commit, more"

stevepiercy commented on docs/narr/advconfig.rst:
```diff
@@ -168,9 +168,12 @@ Using ``config.commit()``
 You can manually commit a configuration by using the
 :meth:`~pyramid.config.Configurator.commit` method between configuration calls.
+After a commit more :term:`configuration declaration`\s may be added to a :term:`configurator`.
+New declarations will not conflict with committed declarations, they will override them.
```
"declarations, but they"

kpinc commented:
Hi Steve,

On Tue, 20 Nov 2018 02:45:26 -0800
Steve Piercy <notifications@github.com> wrote:

> Two minor punctuation and grammar fixes. Otherwise LGTM.

Thanks very much for the help.

I'm not sure about that comma, but ok....   Seems like
you could be right.  Is it really required?

> > @@ -168,9 +168,12 @@ Using ``config.commit()``  
>  
>  You can manually commit a configuration by using the
>  :meth:`~pyramid.config.Configurator.commit` method between
> configuration calls. +After a commit more :term:`configuration
> declaration`\s may be added to a :term:`configurator`. +New
> declarations will not conflict with committed declarations, they will
> override them.
> 
> "declarations, but they"

I think that what happens when there is no longer
a conflict is more memorable without the "but".   The reader
already has it in their head that conflicts are avoided but needs
to know which declarations have "priority" when conflicts are
overridden.  A flat statement without the "but" drives this home.

The comma could be changed to a semi-colon, but
that seems distracting in this case.

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

stevepiercy commented:
The changes I requested must be made to simplify and clarify sentence structure and intent.

I appreciate the stylistic differences you propose, but docs must be accessible to a broad audience. That audience includes people who do not use English as their first language (English language rules are insane), visually impaired users (a comma makes a huge difference to a screen reader, adding a "breath", much like a musical breath symbol), readers who don't have college-level reading comprehension, and folks with reading disabilities.

kpinc commented:
On Tue, 20 Nov 2018 05:03:41 -0800
Steve Piercy <notifications@github.com> wrote:

> The changes I requested must be made to simplify and clarify sentence
> structure and intent.
> 
> I appreciate the stylistic differences you propose, but docs must be
> accessible to a broad audience. That audience includes people who do
> not use English as their first language (English language rules are
> insane), visually impaired users (a comma makes a huge difference to
> a screen reader, adding a "breath", much like a musical breath
> symbol), readers who don't have college-level reading comprehension,
> and folks with reading disabilities.
> 

Humm.  I thought I pushed that comma.

(Is it really wrong to write "After a rain more flowers
may be planted." without a comma after "rain"?)

I disagree about the "but".  I think adding it makes the docs
less clear and less useful.  To make us both happy I've split
the sentence into 2 simpler sentences.

Regards,

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

mmerickel commented on docs/narr/firstapp.rst:
```diff
@@ -190,9 +190,8 @@ WSGI Application Creation
-A call to ``make_wsgi_app`` implies that all configuration is finished
-(meaning all method calls to the configurator, which sets up views and various
-other configuration settings, have been performed).  The ``make_wsgi_app``
+A call to ``make_wsgi_app`` implies that all configuration is finished (meaning all method calls to the configurator, which sets up views and various other configuration :term:`settings`, have been performed).
```
this change doesn't make sense, the settings referred to here are not the literal `settings` object

kpinc commented:
On Wed, 21 Nov 2018 12:37:56 -0800
Michael Merickel <notifications@github.com> wrote:

> mmerickel commented on this pull request.
> 
> 
> 
> > @@ -190,9 +190,8 @@ WSGI Application Creation  
>  
>  After configuring views and ending configuration, the script creates
> a WSGI *application* via
> the :meth:`pyramid.config.Configurator.make_wsgi_app` method. -A call
> to ``make_wsgi_app`` implies that all configuration is finished
> -(meaning all method calls to the configurator, which sets up views
> and various -other configuration settings, have been performed).  The
> ``make_wsgi_app`` +A call to ``make_wsgi_app`` implies that all
> configuration is finished (meaning all method calls to the
> configurator, which sets up views and various other
> configuration :term:`settings`, have been performed).
> 
> this change doesn't make sense, the settings referred to here are not
> the literal `settings` object

Oops.  Reverted.

Thanks.

Karl <kop@meme.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

mmerickel commented:
LGTM thanks @kpinc !

stevepiercy commented:
Thank you!
